File: Wallet.java
import java.util.ArrayList;
import java.util.Scanner;

public class Wallet {
    private double balance;
    private ArrayList<Transaction> transactions;
    private CINInfo cinInfo;
    private String currentCategory;

    public Wallet() {
        this.balance = 0;
        this.transactions = new ArrayList<>();
        this.cinInfo = new CINInfo("", "", "");
        this.currentCategory = "Money";
    }

    public void deposit(double amount) {
        this.balance += amount;
        this.transactions.add(new Transaction("Recharge", amount));
        System.out.println("Dépôt réussi");
    }

    public void withdraw(double amount) {
        if (amount <= this.balance) {
            this.balance -= amount;
            this.transactions.add(new Transaction("Retrait", amount));
            System.out.println("Retrait réussi");
        } else {
            System.out.println("Fonds insuffisants");
        }
    }

    public double getBalance() {
        return this.balance;
    }

    public double getTotalExpenses() {
        double totalExpenses = 0;
        for (Transaction transaction : this.transactions) {
            if (transaction.getType().equals("Retrait")) {
                totalExpenses += transaction.getAmount();
            }
        }
        return totalExpenses;
    }

    public void getTransactionHistory() {
        System.out.println("Suivi des transactions :");
        for (Transaction transaction : this.transactions) {
            System.out.println(transaction.getType() + ": " + transaction.getAmount());
        }
    }

    public void setCINInfo(String nom, String prenom, String numeroCIN) {
        this.cinInfo = new CINInfo(nom, prenom, numeroCIN);
        System.out.println("Informations CIN enregistrées avec succès");
    }

    public void displayCINInfo() {
        System.out.println("Informations CIN :");
        System.out.println("Nom: " + this.cinInfo.getNom());
        System.out.println("Prénom: " + this.cinInfo.getPrenom());
        System.out.println("Numéro CIN: " + this.cinInfo.getNumeroCIN());
    }

    public static void main(String[] args) {
        Wallet myWallet1 = new Wallet();
        Wallet myWallet2 = new Wallet();
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("Menu :");
            System.out.println("1. Money");
            System.out.println("2. CIN");
            System.out.println("0. Quitter");

            System.out.print("Choisissez une catégorie : ");
            String categoryChoice = scanner.nextLine();

            switch (categoryChoice) {
                case "1":
                    myWallet1.currentCategory = "Money";
                    handleMoneyCategory(myWallet1);
                    break;
                case "2":
                    myWallet1.currentCategory = "CIN";
                    handleCINCategory(myWallet1);
                    break;
                case "0":
                    System.out.println("Au revoir !");
                    System.exit(0);
                default:
                    System.out.println("Option invalide. Veuillez choisir une option valide.");
            }
        }
    }

    private static void handleMoneyCategory(Wallet wallet) {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("Money Menu :");
            System.out.println("1. Recharge");
            System.out.println("2. Retrait");
            System.out.println("3. Consultation de solde");
            System.out.println("4. Total des dépenses");
            System.out.println("5. Suivi des transactions");
            System.out.println("0. Retour au menu principal");

            System.out.print("Choisissez une option : ");
            String choice = scanner.nextLine();

            switch (choice) {
                case "1":
                    System.out.print("Entrez le montant à déposer : ");
                    double depositAmount = scanner.nextDouble();
                    wallet.deposit(depositAmount);
                    break;
                case "2":
                    System.out.print("Entrez le montant à retirer : ");
                    double withdrawAmount = scanner.nextDouble();
                    wallet.withdraw(withdrawAmount);
                    break;
                case "3":
                    System.out.println("Solde actuel : " + wallet.getBalance());
                    break;
                case "4":
                    System.out.println("Total des dépenses : " + wallet.getTotalExpenses());
                    break;
                case "5":
                    wallet.getTransactionHistory();
                    break;
                case "0":
                    return;
                default:
                    System.out.println("Option invalide. Veuillez choisir une option valide.");
            }
            scanner.nextLine(); // Consume the newline character left in the buffer
        }
    }

    private static void handleCINCategory(Wallet wallet) {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("CIN Menu :");
            System.out.println("1. Enregistrer les informations CIN");
            System.out.println("2. Afficher les informations CIN");
            System.out.println("0. Retour au menu principal");

            System.out.print("Choisissez une option : ");
            String choice = scanner.nextLine();

            switch (choice) {
                case "1":
                    System.out.print("Entrez votre nom : ");
                    String nom = scanner.nextLine();
                    System.out.print("Entrez votre prénom : ");
                    String prenom = scanner.nextLine();
                    System.out.print("Entrez votre numéro CIN : ");
                    String numeroCIN = scanner.nextLine();
                    wallet.setCINInfo(nom, prenom, numeroCIN);
                    break;
                case "2":
                    wallet.displayCINInfo();
                    break;
                case "0":
                    return;
                default:
                    System.out.println("Option invalide. Veuillez choisir une option valide.");
            }
        }
    }
}

class Transaction {
    private String type;
    private double amount;

    public Transaction(String type, double amount) {
        this.type = type;
        this.amount = amount;
    }

    public String getType() {
        return type;
    }

    public double getAmount() {
        return amount;
    }
}

class CINInfo {
    private String nom;
    private String prenom;
    private String numeroCIN;

    public CINInfo(String nom, String prenom, String numeroCIN) {
        this.nom = nom;
        this.prenom = prenom;
        this.numeroCIN = numeroCIN;
    }

    public String getNom() {
        return nom;
    }

    public String getPrenom() {
        return prenom;
    }

    public String getNumeroCIN() {
        return numeroCIN;
    }
}

File: wallet.php
<?php

class Wallet {
    private $balance;
    private $transactions;
    private $cinInfo;
    private $currentCategory;

    public function __construct() {
        $this->balance = 0;
        $this->transactions = [];
        $this->cinInfo = ['nom' => '', 'prenom' => '', 'numeroCIN' => ''];
        $this->currentCategory = 'Money';
    }

    public function deposit($amount) {
        $this->balance += $amount;
        $this->transactions[] = ['type' => 'Recharge', 'amount' => $amount];
        echo "Dépôt réussi\n";
    }

    public function withdraw($amount) {
        if ($amount <= $this->balance) {
            $this->balance -= $amount;
            $this->transactions[] = ['type' => 'Retrait', 'amount' => $amount];
            echo "Retrait réussi\n";
        } else {
            echo "Fonds insuffisants\n";
        }
    }

    public function getBalance() {
        return $this->balance;
    }

    public function getTotalExpenses() {
        $totalExpenses = 0;
        foreach ($this->transactions as $transaction) {
            if ($transaction['type'] === 'Retrait') {
                $totalExpenses += $transaction['amount'];
            }
        }
        return $totalExpenses;
    }

    public function getTransactionHistory() {
        echo "Suivi des transactions :\n";
        foreach ($this->transactions as $transaction) {
            echo "{$transaction['type']}: {$transaction['amount']}\n";
        }
    }

    public function setCINInfo($nom, $prenom, $numeroCIN) {
        $this->cinInfo = ['nom' => $nom, 'prenom' => $prenom, 'numeroCIN' => $numeroCIN];
        echo "Informations CIN enregistrées avec succès\n";
    }

    public function displayCINInfo() {
        echo "Informations CIN :\n";
        echo "Nom: {$this->cinInfo['nom']}\n";
        echo "Prénom: {$this->cinInfo['prenom']}\n";
        echo "Numéro CIN: {$this->cinInfo['numeroCIN']}\n";
    }
}

$myWallet1 = new Wallet();
$myWallet2 = new Wallet();

while (true) {
    echo "Menu :\n";
    echo "1. Money\n";
    echo "2. CIN\n";
    echo "0. Quitter\n";

    $categoryChoice = readline("Choisissez une catégorie : ");

    switch ($categoryChoice) {
        case '1':
            $myWallet1->currentCategory = 'Money';
            handleMoneyCategory($myWallet1);
            break;
        case '2':
            $myWallet1->currentCategory = 'CIN';
            handleCINCategory($myWallet1);
            break;
        case '0':
            echo "Au revoir !\n";
            exit(0);
        default:
            echo "Option invalide. Veuillez choisir une option valide.\n";
    }
}

function handleMoneyCategory($wallet) {
    while (true) {
        echo "Money Menu :\n";
        echo "1. Recharge\n";
        echo "2. Retrait\n";
        echo "3. Consultation de solde\n";
        echo "4. Total des dépenses\n";
        echo "5. Suivi des transactions\n";
        echo "0. Retour au menu principal\n";

        $choice = readline("Choisissez une option : ");

        switch ($choice) {
            case '1':
                $depositAmount = readline("Entrez le montant à déposer : ");
                $wallet->deposit((float)$depositAmount);
                break;
            case '2':
                $withdrawAmount = readline("Entrez le montant à retirer : ");
                $wallet->withdraw((float)$withdrawAmount);
                break;
            case '3':
                echo "Solde actuel : " . $wallet->getBalance() . "\n";
                break;
            case '4':
                echo "Total des dépenses : " . $wallet->getTotalExpenses() . "\n";
                break;
            case '5':
                $wallet->getTransactionHistory();
                break;
            case '0':
                return;
            default:
                echo "Option invalide. Veuillez choisir une option valide.\n";
        }
    }
}

function handleCINCategory($wallet) {
    while (true) {
        echo "CIN Menu :\n";
        echo "1. Enregistrer les informations CIN\n";
        echo "2. Afficher les informations CIN\n";
        echo "0. Retour au menu principal\n";

        $choice = readline("Choisissez une option : ");

        switch ($choice) {
            case '1':
                $nom = readline("Entrez votre nom : ");
                $prenom = readline("Entrez votre prénom : ");
                $numeroCIN = readline("Entrez votre numéro CIN : ");
                $wallet->setCINInfo($nom, $prenom, $numeroCIN);
                break;
            case '2':
                $wallet->displayCINInfo();
                break;
            case '0':
                return;
            default:
                echo "Option invalide. Veuillez choisir une option valide.\n";
        }
    }
}
